trojsten/rules/susi.py

```diff
@@ -0,0 +1,188 @@
+            EventParticipant.objects.filter(
+                Q(
+                    # TODO I think commenting next line will retrieve participations in
+                    # all trojsten camps
+                    # event__type__name=SUSI_CAMP_TYPE,
+                    event__end_time__lt=round.end_time,
+                    event__end_time__year__gte=round.end_time.year - SUSI_YEARS_OF_CAMPS_HISTORY,
+                ),
+                Q(going=True) | Q(type=EventParticipant.PARTICIPANT),
+            )
+            .values("user")
+            .annotate(camps=Count("event__semester", distinct=True))
+            .values_list("user", "camps")
+        )
+
+        # We ignore camps that happened before SUSI_YEARS_OF_CAMPS_HISTORY years, so we don't
+        # produce too big dictionaries of users.
+        self.susi_camps = dict(
+            EventParticipant.objects.filter(
+                Q(
+                    event__type__name=SUSI_CAMP_TYPE,
+                    event__end_time__lt=round.end_time,
+                    event__end_time__year__gte=round.end_time.year - SUSI_YEARS_OF_CAMPS_HISTORY,
+                ),
+                Q(going=True) | Q(type=EventParticipant.PARTICIPANT),
+            )
+            .values("user")
+            .annotate(camps=Count("event__semester", distinct=True))
+            .values_list("user", "camps")
+        )
+
+        # TODO Fill dictionary with values of property "SUSI_Ucasti_na_sifrovackach" for all
+        # susi participants
+        self.puzzlehunt_participations = None
+
+        # TODO Fill dictionary with booleans, indicating whether a participant has ended
+        #  high school or not
+        self.has_graduated = None
+
+    def run(self, res_request):
+        self.prepare_coefficients(res_request.round)
+        res_request.has_submit_in_blyskavica = set()
+        for submit in Submit.objects.filter(
+            task__round__semester=res_request.round.semester, task__number__in=[6, 7]
+        ).select_related("user"):
+            res_request.has_submit_in_blyskavica.add(submit.user)
+        return super(SUSIResultsGenerator, self).run(res_request)
+
+    def is_user_active(self, request, user):
+        active = super(SUSIResultsGenerator, self).is_user_active(request, user)
+        coefficient = self.get_user_coefficient(user, request.round)
+
+        if self.tag.key == SUSI_AGAT:
+            active = active and (
+                (coefficient <= SUSI_AGAT_MAX_COEFFICIENT)
+                and not self.get_graduation_status(user, request.round)
+            )
+
+        if self.tag.key == SUSI_BLYSKAVICA:
+            active = active and (
+                (
+                    coefficient > SUSI_AGAT_MAX_COEFFICIENT
+                    or user in request.has_submit_in_blyskavica
+                )
+                and not (self.get_graduation_status(user, request.round))
+            )
+
+        if self.tag.key == SUSI_CIFERSKY_CECH:
```

Comment:
Robí toto niečo?

Reply:
tbh neviem, ale kms to malo tak som to ukradol